### PR TITLE
hostap: add 11v BTM query support

### PR DESCRIPTION
**Pull request by Rex-Chen-NXP**
Add 11v BTM query L2 layer cmd support.

### CODE REVIEW

File: include/zephyr/net/wifi.h [hunk 1]
```diff
@@ -484,6 +484,16 @@ static const char * const wifi_ps_param_config_err_code_tbl[] = {
 };
 /** @endcond */
 
+/** 11v BSS transition management Query reasons. */
+enum wifi_btm_query_reason {
+	/** Unspecified. */
+	WIFI_BTM_QUERY_REASON_UNDPECIFIED = 0,
```
Comment (krish2718) [inline include/zephyr/net/wifi.h:490]: please add some references for these reasons 0/16/20.

File: include/zephyr/net/wifi.h [hunk 2]
```diff
@@ -484,6 +484,16 @@ static const char * const wifi_ps_param_config_err_code_tbl[] = {
 };
 /** @endcond */
 
+/** 11v BSS transition management Query reasons. */
```
Comment (krish2718) [inline include/zephyr/net/wifi.h:487]:
```suggestion
/** IEEE 802.11v BTM (BSS transition management) Query reasons. */
```

File: include/zephyr/net/wifi_mgmt.h [hunk 3]
```diff
@@ -920,6 +928,15 @@ struct wifi_mgmt_ops {
 	 * @return 0 if ok, < 0 if error
 	 */
 	int (*channel)(const struct device *dev, struct wifi_channel_info *channel);
+	/** Send bss transition query
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:931]:
```suggestion
	/** Send BTM query
```

File: modules/hostap/src/supp_api.c [hunk 4]
```diff
@@ -905,6 +905,32 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
...
+		goto out;
+	}
+
+	ret = 0;
+
+out:
+	k_mutex_unlock(&wpa_supplicant_mutex);
+
+	return ret;
+}
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:932]: protect with `WIFI_NM_WPA_SUPPLICANT_WNM` including the assignment. WNM can be disabled to save memory.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 5]
```diff
@@ -1543,6 +1543,30 @@ static int cmd_wifi_listen_interval(const struct shell *sh, size_t argc, char *a
...
+static int cmd_wifi_btm_query(const struct shell *sh, size_t argc, char *argv[])
+{
+	struct net_if *iface = net_if_get_first_wifi();
+	uint8_t query_reason = 0;
+
+	context.sh = sh;
+
+	if (!parse_number(sh, (long *)&query_reason, argv[1],
+			  WIFI_BTM_QUERY_REASON_UNDPECIFIED,
+			  WIFI_BTM_QUERY_REASON_LOW_RSSI)) {
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1555]:
```suggestion
			  WIFI_BTM_QUERY_REASON_LEAVING_ESS)) {
```

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 6]
```diff
@@ -754,6 +754,22 @@ static int wifi_get_version(uint32_t mgmt_request, struct net_if *iface,
...
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
+	uint8_t query_reason = *((uint8_t *)data);
+
+	if (wifi_mgmt_api == NULL || wifi_mgmt_api->btm_query == NULL) {
+		return -ENOTSUP;
+	}
+
+	return wifi_mgmt_api->btm_query(dev, query_reason);
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_mgmt.c:768]: should also check for valid reason (based on enum) before calling driver API? or do you want allow arbitrary reasons and let implementation handle this?

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: modules/hostap/Kconfig [hunk 7]
```diff
@@ -395,6 +395,7 @@ config WEP
 
 config WPA_CLI
 	bool
+	default y
```
Comment (krish2718) [inline modules/hostap/Kconfig:398]: I guess this is accidental?

Comment (Rex-Chen-NXP) [reply modules/hostap/Kconfig:398]: > I guess this is accidental?

Removed this change

File: include/zephyr/net/wifi.h [hunk 8]
```diff
@@ -484,6 +484,20 @@ static const char * const wifi_ps_param_config_err_code_tbl[] = {
 };
 /** @endcond */
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_WNM
+/* IEEE 802.11v BTM (BSS transition management) Query reasons. 
+ * Refer to IEEE Std 802.11v - Table 7-43x-Transition and Transition Query reasons table.
+ */
```
[the remark this answers is not in the document]
Comment (krish2718) [reply include/zephyr/net/wifi.h:490]: I can see a trailing whitespace in line#488 and also the comment should start with `/**`

File: include/zephyr/net/wifi.h [hunk 9]
```diff
@@ -484,6 +484,20 @@ static const char * const wifi_ps_param_config_err_code_tbl[] = {
 };
 /** @endcond */
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_WNM
+/** IEEE 802.11v BTM (BSS transition management) Query reasons.
+ * Refer to IEEE Std 802.11v - Table 7-43x-Transition and Transition Query reasons table.
+ */
+enum wifi_btm_query_reason {
+	/** Unspecified. */
+	WIFI_BTM_QUERY_REASON_UNDPECIFIED = 0,
```
Comment (krish2718) [inline include/zephyr/net/wifi.h:493]:
```suggestion
	WIFI_BTM_QUERY_REASON_UNSPECIFIED = 0,
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 10]
```diff
@@ -1543,6 +1543,32 @@ static int cmd_wifi_listen_interval(const struct shell *sh, size_t argc, char *a
...
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_WNM
+static int cmd_wifi_btm_query(const struct shell *sh, size_t argc, char *argv[])
+{
+	struct net_if *iface = net_if_get_first_wifi();
+	uint8_t query_reason = 0;
+
+	context.sh = sh;
+
+	if (!parse_number(sh, (long *)&query_reason, argv[1], NULL,
+			  WIFI_BTM_QUERY_REASON_UNDPECIFIED,
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1555]:
```suggestion
			  WIFI_BTM_QUERY_REASON_UNSPECIFIED,
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 11]
```diff
@@ -1543,6 +1543,32 @@ static int cmd_wifi_listen_interval(const struct shell *sh, size_t argc, char *a
...
+			  WIFI_BTM_QUERY_REASON_LEAVING_ESS)) {
+		return -EINVAL;
+	}
+
+	if (net_mgmt(NET_REQUEST_WIFI_BTM_QUERY, iface,
+			  &query_reason, sizeof(query_reason))) {
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1561]: alignment seems off

File: subsys/net/l2/wifi/wifi_mgmt.c (hunk 6 above)
Comment (krish2718) [reply subsys/net/l2/wifi/wifi_mgmt.c:769]: If you move the check from shell->mgmt then it covers both shell and API cases. Also, as said above check covers the full range min-max, but we only have three enumerations, so, need to extend check or add a enumeration for UNKNOWN_REASON and encode anyother values as UNKNOWN_REASON.

File: include/zephyr/net/wifi.h [hunk 12]
```diff
@@ -484,6 +484,20 @@ static const char * const wifi_ps_param_config_err_code_tbl[] = {
 };
 /** @endcond */
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_WNM
+/** IEEE 802.11v BTM (BSS transition management) Query reasons.
+ * Refer to IEEE Std 802.11v - Table 7-43x-Transition and Transition Query reasons table.
```
Comment (krish2718) [inline include/zephyr/net/wifi.h:489]: please add specific year, as sections can change and better to use 802.11-2020 and later versions

File: include/zephyr/net/wifi.h (hunk 9 above)
Comment (Rex-Chen-NXP) [reply include/zephyr/net/wifi.h]: Updated as comment

File: subsys/net/l2/wifi/wifi_shell.c (hunk 10 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated as comment

File: include/zephyr/net/wifi.h (hunk 12 above)
Comment (Rex-Chen-NXP) [reply include/zephyr/net/wifi.h]: Updated as comment

File: subsys/net/l2/wifi/wifi_mgmt.c (hunk 6 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_mgmt.c]: Updated as comment

File: subsys/net/l2/wifi/wifi_shell.c (hunk 11 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 13]
```diff
@@ -754,6 +754,29 @@ static int wifi_get_version(uint32_t mgmt_request, struct net_if *iface,
...
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
+	uint8_t query_reason = *((uint8_t *)data);
+
+	if (wifi_mgmt_api == NULL || wifi_mgmt_api->btm_query == NULL) {
+		return -ENOTSUP;
+	}
+
+	if (query_reason >= WIFI_BTM_QUERY_REASON_UNSPECIFIED &&
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_mgmt.c:769]: Should we limit to specific reasons?

```
	if (query_reason != WIFI_BTM_QUERY_REASON_UNSPECIFIED ||
		query_reason !=WIFI_BTM_QUERY_REASON_LOW_RSSI ||
		query_reason != WIFI_BTM_QUERY_REASON_LEAVING_ESS) {
		return -EINVAL;
	}
```

**Review by krish2718 — DISMISSED** [review]

File: subsys/net/l2/wifi/wifi_mgmt.c (hunk 13 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_mgmt.c:784]: This logic is to limit input parameter: 
if (query_reason >= WIFI_BTM_QUERY_REASON_UNSPECIFIED &&
	query_reason <= WIFI_BTM_QUERY_REASON_LEAVING_ESS)

Comment (krish2718) [reply subsys/net/l2/wifi/wifi_mgmt.c:784]: it's not really a big deal, but you are allowing the full range from `0-20` while valid reasons are only `0,16,20`

Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_mgmt.c:784]: We only enum 0,16,20 here, but input reason from 0-20 is correct and acceptable for supplicant

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 14]
```diff
@@ -754,6 +754,29 @@ static int wifi_get_version(uint32_t mgmt_request, struct net_if *iface,
...
+		query_reason <= WIFI_BTM_QUERY_REASON_LEAVING_ESS) {
```
[the remark this answers is not in the document]
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_mgmt.c]: Added tab before this new line

[the remark this answers is not in the document]
Comment (krish2718) [reply subsys/net/l2/wifi/wifi_mgmt.c]: As the compliance suggests better to run `clang-format` on this change, one without your changes to fix existing formatting errors and one with yours.

**Review by krish2718 — DISMISSED** [review]

**Review by jukkar — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]
Please fix the formatting issues reported

**Comment by Rex-Chen-NXP** [comment]
Have formatted all new add code

**Comment by Rex-Chen-NXP** [comment]
Hi @jukkar  @krish2718 
Could you help merge pr?

**Comment by krish2718** [comment]
> Have formatted all new add code

I still see issues being reported by `clang-format` https://github.com/zephyrproject-rtos/zephyr/actions/runs/10402786647/job/28807970603?pr=76251 (also in the `files changed` tab too), though it's optional good to fix these.

**Comment by Rex-Chen-NXP** [comment]
> > Have formatted all new add code
> 
> I still see issues being reported by `clang-format` https://github.com/zephyrproject-rtos/zephyr/actions/runs/10402786647/job/28807970603?pr=76251 (also in the `files changed` tab too), though it's optional good to fix these.

I didn't see any check failed from CI result of this pr and also your shared linker, which piece of code should I change?

**Comment by krish2718** [comment]
> I didn't see any check failed from CI result of this pr

CI will still pass, as clang-format is optional

> also your shared linker, which piece of code should I change?

Do you see the below in the `files changed` tab? The link shows the same as well.

image
image

**Pull request merged**